Failed to install python-lsp-server

### I've searched open issues for similar requests

Yes

### I've recently downloaded the latest plugin version of mason.nvim

- [X] Yes

### Problem description

Hey,
When I try to install python-lsp-server the installation process is failing with the attached error.

I have managed to recreate the issue. When I manually create a venv using the command: `python3.8 -m venv venv` On ubuntu 18.04, the pip version inside the venv is 9.0.1, which is ancient. The venv module itself was installed with: `sudo apt install python3.8-venv`.
Also, I have set the `python3_host_prog` to `python3.8` because the system's default is 3.6 and pylsp requires minimum version 3.7.

When I create the venv but before installing pylsp I upgrade the pip (`pip install -U pip`), the problem disappears.
I would be happy to open a merge request and fix it myself, but I'm not sure what is the best and cleanest way to solve it.

### Expected behavior

Successful installtion of `python-lsp-server` package.

### Affected packages

python-lsp-server

### Mason output

```Text
Failed to build ujson
Installing collected packages: parso, jedi, ujson, python-lsp-jsonrpc, setuptools, pluggy, pyflakes, pycodestyle, mccabe, flake8, toml, autopep8, whatthepatch, pyparsing, packaging, tomli, pytoolconfig, rope, snowballstemmer, pydocstyle, dill, tomlkit, typing-extensions, lazy-object-proxy, wrapt, astroid, isort, platformdirs, pylint, yapf, python-lsp-server
  Running setup.py install for ujson: started
    Running setup.py install for ujson: finished with status 'error'
    Complete output from command /home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-ek_ql5jf/ujson/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-zu4hu9as-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include/site/python3.8/ujson:
    /tmp/pip-build-ek_ql5jf/ujson/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/integration.py:27: RuntimeWarning:
    ERROR: setuptools==39.0.1 is used in combination with setuptools_scm>=6.x
    
    Your build configuration is incomplete and previously worked by accident!
    setuptools_scm requires setuptools>=45
    
    
    This happens as setuptools is unable to replace itself when a activated build dependency
    requires a more recent setuptools version
    (it does not respect "setuptools>X" in setup_requires).
    
    
    setuptools>=31 is required for setup.cfg metadata support
    setuptools>=42 is required for pyproject.toml configuration support
    
    Suggested workarounds if applicable:
     - preinstalling build dependencies like setuptools_scm before running setup.py
     - installing setuptools_scm using the system package manager to ensure consistency
     - migrating from the deprecated setup_requires mechanism to pep517/518
       and using a pyproject.toml to declare build dependencies
       which are reliably pre-installed before running the build tools
    
      warnings.warn(
    running install
    running build
    running build_ext
    building 'ujson' extension
    creating build
    creating build/temp.linux-x86_64-3.8
    creating build/temp.linux-x86_64-3.8/deps
    creating build/temp.linux-x86_64-3.8/deps/double-conversion
    creating build/temp.linux-x86_64-3.8/deps/double-conversion/double-conversion
    creating build/temp.linux-x86_64-3.8/lib
    creating build/temp.linux-x86_64-3.8/python
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/strtod.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/strtod.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/double-conversion.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/double-conversion.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/diy-fp.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/diy-fp.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/bignum.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/bignum.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/bignum-dtoa.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/bignum-dtoa.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/fast-dtoa.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/fast-dtoa.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/fixed-dtoa.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/fixed-dtoa.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./deps/double-conversion/double-conversion/cached-powers.cc -o build/temp.linux-x86_64-3.8/./deps/double-conversion/double-conversion/cached-powers.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./lib/dconv_wrapper.cc -o build/temp.linux-x86_64-3.8/./lib/dconv_wrapper.o -D_GNU_SOURCE
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./python -I./lib -I./deps/double-conversion/double-conversion -I/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include -I/usr/include/python3.8 -c ./python/ujson.c -o build/temp.linux-x86_64-3.8/./python/ujson.o -D_GNU_SOURCE
    ./python/ujson.c:39:10: fatal error: Python.h: No such file or directory
     #include <Python.h>
              ^~~~~~~~~~
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    
    ----------------------------------------
Command "/home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-ek_ql5jf/ujson/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-zu4hu9as-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/omer/projects/config/neovim/etc/xdg/nvim/mason/packages/python-lsp-server/venv/include/site/python3.8/ujson" failed with error code 1 in /tmp/pip-build-ek_ql5jf/ujson/
spawn: python failed with exit code 1 and signal 0.
```


### Installation log

```Text
The log is too big, so I attached it in the screenshot section :)
```


### Neovim version (>= 0.7)

NVIM v0.8.0-v0.8.0
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-10 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim-neovim/neovim-neovim/build/cmake.config -I/home/runner/work/neovim-neovim/neovim-neovim/src -I/home/runner/work/neovim-neovim/neovim-neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim-neovim/neovim-neovim/build/src/nvim/auto -I/home/runner/work/neovim-neovim/neovim-neovim/build/include
Compiled by runner@fv-az162-912

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/runner/work/neovim-neovim/neovim-neovim/build/nvim.AppDir/usr/share/nvim
"

Run :checkhealth for more info

### Operating system/version

Linux <hostname> 5.4.0-131-generic #147~18.04.1-Ubuntu SMP Sat Oct 15 13:10:18 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Healthcheck

```Text
mason: require("mason.health").check()                                                                                                                                                                               
========================================================================                                                                                                                                             
## mason.nvim report                                                                                                                                                                                                 
  - OK: neovim version >= 0.7.0                                                                                                                                                                                      
  - WARNING: Go: not available                                                                                                                                                                                       
  - WARNING: cargo: not available                                                                                                                                                                                    
  - WARNING: luarocks: not available                                                                                                                                                                                 
  - OK: Ruby: `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]`                                                                                                                                         
  - OK: RubyGem: `2.7.6`                                                                                                                                                                                             
  - WARNING: Composer: not available                                                                                                                                                                                 
  - WARNING: PHP: not available                                                                                                                                                                                      
  - ERROR: npm: not available                                                                                                                                                                                        
  - ERROR: node: not available                                                                                                                                                                                       
  - OK: python3: `Python 3.6.9`                                                                                                                                                                                      
  - OK: pip3: `pip 21.3.1 from /usr/local/lib/python3.6/dist-packages/pip (python 3.6)`                                                                                                                              
  - WARNING: javac: not available                                                                                                                                                                                    
  - OK: java: `openjdk version "11.0.16" 2022-07-19`                                                                                                                                                                 
  - WARNING: julia: not available                                                                                                                                                                                    
  - OK: wget: `GNU Wget 1.19.4 built on linux-gnu.`                                                                                                                                                                  
  - OK: curl: `curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.1 zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3`                                                   
  - OK: gzip: `gzip 1.6`                                                                                                                                                                                             
  - OK: tar: `tar (GNU tar) 1.29`                                                                                                                                                                                    
  - OK: bash: `GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)`                                                                                                                                            
  - OK: sh: `Ok`                                                                                                                                                                                                     
  - OK: python3_host_prog: `Python 3.8.0`                                                                                                                                                                            
  - OK: GitHub API rate limit. Used: 0. Remaining: 60. Limit: 60. Reset: Tue 01 Nov 2022 23:23:37 IST.
```


### Screenshots

[mason.log](https://github.com/williamboman/mason.nvim/files/9913790/mason.log)

--- a/README.md
+++ b/README.md
@@ -163,6 +163,9 @@ local DEFAULT_SETTINGS = {
     PATH = "prepend",
 
     pip = {
+        -- Whether to upgrade pip to the latest version in the virtual environment before installing packages.
+        upgrade_pip = false,
+
         -- These args will be added to `pip install` calls. Note that setting extra args might impact intended behavior
         -- and is not recommended.
         --

--- a/doc/mason.txt
+++ b/doc/mason.txt
@@ -215,6 +215,9 @@ Example:
         PATH = "prepend",
 
         pip = {
+            -- Whether to upgrade pip to the latest version in the virtual environment before installing packages.
+            upgrade_pip = false,
+
             -- These args will be added to `pip install` calls. Note that setting extra args might impact intended behavior
             -- and is not recommended.
             --

--- a/lua/mason-core/managers/pip3/init.lua
+++ b/lua/mason-core/managers/pip3/init.lua
@@ -64,6 +64,18 @@ function M.install(packages)
 
     Optional.of_nilable(executable)
         :if_present(function()
+            if settings.current.pip.upgrade_pip then
+                ctx.spawn.python {
+                    "-m",
+                    "pip",
+                    "--disable-pip-version-check",
+                    "install",
+                    "-U",
+                    settings.current.pip.install_args,
+                    "pip",
+                    with_paths = { M.venv_path(ctx.cwd:get()) },
+                }
+            end
             ctx.spawn.python {
                 "-m",
                 "pip",

--- a/lua/mason/settings.lua
+++ b/lua/mason/settings.lua
@@ -15,6 +15,9 @@ local DEFAULT_SETTINGS = {
     PATH = "prepend",
 
     pip = {
+        -- Whether to upgrade pip to the latest version in the virtual environment before installing packages.
+        upgrade_pip = false,
+
         -- These args will be added to `pip install` calls. Note that setting extra args might impact intended behavior
         -- and is not recommended.
         --
@@ -92,7 +95,7 @@ M.current = M._DEFAULT_SETTINGS
 
 ---@param opts MasonSettings
 function M.set(opts)
-    M.current = vim.tbl_deep_extend("force", M.current, opts)
+    M.current = vim.tbl_deep_extend("force", vim.deepcopy(M.current), opts)
 end
 
 return M

--- a/tests/mason-core/managers/pip3_spec.lua
+++ b/tests/mason-core/managers/pip3_spec.lua
@@ -13,6 +13,10 @@ local spawn = require "mason-core.spawn"
 local api = require "mason-registry.api"
 
 describe("pip3 manager", function()
+    before_each(function()
+        settings.set(settings._DEFAULT_SETTINGS)
+    end)
+
     it("normalizes pip3 packages", function()
         local normalize = pip3.normalize_package
         assert.equals("python-lsp-server", normalize "python-lsp-server[all]")
@@ -124,6 +128,41 @@ describe("pip3 manager", function()
                 "install",
                 "-U",
                 { "--proxy", "http://localhost:8080" },
+                { "package" },
+                with_paths = { path.concat { path.package_prefix "dummy", "venv", "bin" } },
+            }
+        end)
+    )
+
+    it(
+        "should upgrade pip",
+        async_test(function()
+            settings.set {
+                pip = {
+                    upgrade_pip = true,
+                },
+            }
+            local handle = InstallHandleGenerator "dummy"
+            local ctx = InstallContextGenerator(handle)
+            installer.run_installer(ctx, pip3.packages { "package" })
+            vim.pretty_print(ctx.spawn.python)
+            assert.spy(ctx.spawn.python).was_called_with {
+                "-m",
+                "pip",
+                "--disable-pip-version-check",
+                "install",
+                "-U",
+                {},
+                "pip" ,
+                with_paths = { path.concat { path.package_prefix "dummy", "venv", "bin" } },
+            }
+            assert.spy(ctx.spawn.python).was_called_with {
+                "-m",
+                "pip",
+                "--disable-pip-version-check",
+                "install",
+                "-U",
+                {},
                 { "package" },
                 with_paths = { path.concat { path.package_prefix "dummy", "venv", "bin" } },
             }

--- a/tests/mason-core/managers/pip3_spec.lua
+++ b/tests/mason-core/managers/pip3_spec.lua
@@ -153,7 +153,7 @@ describe("pip3 manager", function()
                 "install",
                 "-U",
                 {},
-                "pip" ,
+                "pip",
                 with_paths = { path.concat { path.package_prefix "dummy", "venv", "bin" } },
             }
             assert.spy(ctx.spawn.python).was_called_with {

--- a/tests/mason-core/managers/pip3_spec.lua
+++ b/tests/mason-core/managers/pip3_spec.lua
@@ -121,6 +121,7 @@ describe("pip3 manager", function()
             local handle = InstallHandleGenerator "dummy"
             local ctx = InstallContextGenerator(handle)
             installer.run_installer(ctx, pip3.packages { "package" })
+            assert.spy(ctx.spawn.python).was_called(1)
             assert.spy(ctx.spawn.python).was_called_with {
                 "-m",
                 "pip",
@@ -146,6 +147,7 @@ describe("pip3 manager", function()
             local ctx = InstallContextGenerator(handle)
             installer.run_installer(ctx, pip3.packages { "package" })
             vim.pretty_print(ctx.spawn.python)
+            assert.spy(ctx.spawn.python).was_called(2)
             assert.spy(ctx.spawn.python).was_called_with {
                 "-m",
                 "pip",

